packages/block-editor/src/components/rich-text/index.js

```diff
@@ -459,12 +470,18 @@ export class RichText extends Component {
 	 * Handles the `selectionchange` event: sync the selection to local state.
 	 */
 	onSelectionChange() {
-		const value = this.createRecord();
-		const { start, end } = value;
+		const { start, end } = this.createRecord();
 
 		if ( start !== this.state.start || end !== this.state.end ) {
 			const { isCaretWithinFormattedText } = this.props;
-			const activeFormats = getActiveFormats( value );
+			const value = this.getRecord();
+			const newValue = {
+				...value,
+				start,
+				end,
+			};
+			delete newValue.activeFormats;
+			const activeFormats = getActiveFormats( newValue );
```

Comment:
If we're working with a shallow clone, seems we might as well assign `activeFormats` directly, rather than create another shallow clone below?

```js
const newValue = {
	...value,
	activeFormats: getActiveFormats( newValue ),
	start,
	end,
};
```